keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1077 @@
+tightly scoped this proposal, and hopefully struck an effective balance between configurability and
+usefulness that still leaves plenty of room for 3rd party extensions. We are also providing a
+library implementation and spec to encourage custom controller development.
+
+**Exemptions creep:** We have already gotten requests to trigger exemptions on various different
+fields. The more exemption knobs we add, the harder it becomes to comprehend the current state of
+the policy, and the more room there is for error. To prevent this, we should be very conservative
+about adding new exemption dimensions. The existing knobs were carefully chosen with specific
+extensibility use cases in mind.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Updates
+
+Updates to the following pod fields are exempt from policy checks, meaning that if a pod update
+request only changes these fields it will not be denied even if the pod is in violation of the
+current policy level:
+
+- Any metadata updates _EXCEPT_ changes to the seccomp or apparmor annotations:
+    - `seccomp.security.alpha.kubernetes.io/pod` (deprecated)
+    - `container.seccomp.security.alpha.kubernetes.io/*` (deprecated)
+    - `container.apparmor.security.beta.kubernetes.io/*`
+- Valid updates to `.spec.activeDeadlineSeconds`
+- Valid updates to `.spec.tolerations`
+- Valid updates to [Pod resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources)
+
+Note that updates to container images WILL require a policy reevaluation.
+
+Pod status & nodeName updates are handled by `status` and `binding` subresource requests
+respectively, and are not checked against policies.
+
+Update requests to Pods and [PodTemplate resources](#podtemplate-resources) will reevaluate the full object
+against audit & warn policies, independent of which fields are being modified.
+
+#### Ephemeral Containers
+
+In the initial implementation, ephemeral containers will be subject to the same policy restrictions,
+and adding or updating ephemeral containers will require a full policy check.
+
+<<[UNRESOLVED]>>
+
+_Non-blocking for alpha. This should be resolved for beta._
+
+Once ephemeral containers allow [custom security contexts], it may be desirable to run an ephemeral
+container with higher privileges for debugging purposes. For example, CAP_SYS_PTRACE is forbidden by
+the baseline policy but can be useful in debugging. We could introduce yet-another-mode-label that
+only applies enforcement to ephemeral containers (defaults to the enforce policy).
```

Comment:
I don't think it makes sense to allow ephemeral containers that exceed the configured policy level for the namespace. If you want ephemeral containers to be able to use something like CAP_SYS_PTRACE, the policy level for the namespace must be one that allows CAP_SYS_PTRACE.

That doesn't mean pods would *have* to have CAP_SYS_PTRACE, just that they would be permitted to by policy if desired. You can always create a more restricted pod inside a baseline or privileged namespace if you want.

Reply:
I like the idea of having some mechanism that further locks down what an ephemeral container can do. Such a mechanism feels like it would be well out of scope for _this_ KEP.

Reply:
I think it depends. If you're looking at a hard-multitenancy environment where you want tenants to be completely isolated from each other, than absolutely - tenants should not be able to escalate privilege through ephemeral containers.

However, if the primary concern is limiting what an external attacker could do by compromising a production workload, then I think the ability to temporarily escalate privilege in the ephemeral container would be valuable for debugging. Also, if we provide an escape hatch we leave the door open for a 3rd party controller to manage more fine-grained control over that privilege escalation (e.g. you must be on-call or have an issue ticket attached to the request).

Reply:
> if the primary concern is limiting what an external attacker could do by compromising a production workload, then I think the ability to temporarily escalate privilege in the ephemeral container would be valuable for debugging

That doesn't make sense to me. Compromising a production workload means they gain control of the process inside the pod, not necessarily control over the Pod API surface.

If you want to temporarily escalate permissions on a container running in a namespace, I'd expect a temporary policy change on the namespace, not exempting ephemeral containers from this policy.

Reply:
I would not expect to have different rules for various types of containers inside of a pod.  I expect the namespace labels to be effective against all current and future types of containers to avoid having certain containers bypass the security constraints.